packages/wallet-sdk/src/connection/WalletLinkConnection.ts

```diff
@@ -101,111 +73,98 @@ export class WalletLinkConnection {
-        .pipe(
-          tap(
-            (state) =>
-              this.diagnostic?.log(EVENTS.CONNECTED_STATE_CHANGE, {
-                state,
-                sessionIdHash: Session.hash(sessionId),
-              })
-          ),
-          // ignore initial DISCONNECTED state
-          skip(1),
+    this.ws.setConnectionStateListener(async (state) => {
+      // attempt to reconnect every 5 seconds when disconnected
+      this.diagnostic?.log(EVENTS.CONNECTED_STATE_CHANGE, {
+        state,
+        sessionIdHash: Session.hash(sessionId),
+      });
+
+      let connected = false;
+      switch (state) {
+        case ConnectionState.DISCONNECTED:
           // if DISCONNECTED and not destroyed
-          filter((cs) => cs === ConnectionState.DISCONNECTED && !this.destroyed),
-          // wait 5 seconds
-          delay(5000),
-          // check whether it's destroyed again
-          filter((_) => !this.destroyed),
-          // reconnect
-          flatMap((_) => ws.connect()),
-          retry()
-        )
-        .subscribe()
-    );
-
-    // perform authentication upon connection
-    this.subscriptions.add(
-      ws.connectionState$
-        .pipe(
-          // ignore initial DISCONNECTED and CONNECTING states
-          skip(2),
-          switchMap((cs) =>
-            iif(
-              () => cs === ConnectionState.CONNECTED,
-              // if CONNECTED, authenticate, and then check link status
-              this.authenticate().pipe(
-                tap((_) => this.sendIsLinked()),
-                tap((_) => this.sendGetSessionConfig()),
-                map((_) => true)
-              ),
-              // if not CONNECTED, emit false immediately
-              of(false)
-            )
-          ),
-          distinctUntilChanged(),
-          catchError((_) => of(false))
-        )
-        .subscribe((connected) => {
-          this.connectedSubject.next(connected);
-          this.connectedListener?.(connected);
-        })
-    );
-
-    // send heartbeat every n seconds while connected
-    this.subscriptions.add(
-      ws.connectionState$
-        .pipe(
-          // ignore initial DISCONNECTED state
-          skip(1),
-          switchMap((cs) =>
-            iif(
-              () => cs === ConnectionState.CONNECTED,
-              // if CONNECTED, start the heartbeat timer
-              timer(0, HEARTBEAT_INTERVAL)
-            )
-          )
-        )
-        .subscribe((i) =>
+          if (!this.destroyed) {
+            const connect = async () => {
+              // wait 5 seconds
+              await new Promise((resolve) => setTimeout(resolve, 5000));
+              // check whether it's destroyed again
+              if (!this.destroyed) {
+                // reconnect
+                ws.connect().catch(() => {
+                  connect();
+                });
+              }
+            };
+            connect();
+          }
+          break;
+
+        case ConnectionState.CONNECTED:
```

Comment:
We were skipping 2 events before authenticating.  Not sure why this was here before but something to verify -are we sure we're not calling authenticate twice?

Reply:
for `ConnectionState.CONNECTED` check which used to have `skip(2)` , i believe it was to ignore initial `DISCONNECTED` and `CONNECTING` states as the comment said. we don’t need this skipping anymore for this code block as we moved to simple switch syntax instead of individual rx subscription.